storage: MVCC range tombstones feature branch

erikgrinaker opened the pull request:
This is a long-lived feature branch for MVCC range tombstones. It will be used until 22.1 is released, to ease backporting during stabilization. Merged PRs:

* #79291
* #79744

erikgrinaker committed:
storage: upgrade to pebblev2 table format

The `Pebblev2` SSTable format adds support for range keys. Add two new
cluster versions to provide the upgrade path - the first version for
bumping the store, the second for use as a feature gate.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77384)
<!-- Reviewable:end -->

erikgrinaker committed:
vendor: bump Pebble

Release note: None

erikgrinaker committed:
storage: remove `MVCCIterator.SetUpperBound` and tweak `ClearIterRange`

This patch removes the `MVCCIterator.SetUpperBound()` method, which was
only used by `MVCCIterator.ClearIterRange()`.

`ClearIterRange` now also creates the iterator itself, instead of having
it passed in. This ensures the iterator is created with appropriate
bounds and options, which becomes even more important with the
introduction of range keys. If further flexibility is needed, we can
consider passing an `IterOptions` struct instead of a key span.

Release note: None

erikgrinaker committed:
storage: use `SetOptions()` when reusing Pebble iterators

Iterator reuse now relies on `Pebble.SetOptions()` to configure the
reused Pebble iterator. This allows a wider range of iterators to be
reused, since previously only the bounds could be changed on existing
iterators.

Release note: None

erikgrinaker committed:
storage: handle multiple iterators in `pebbleReadOnly` and `pebbleBatch`

Previously, `pebbleReadOnly` and `pebbleBatch` only supported a single
concurrent iterator of a given type. Attempts to create an additional
iterator would panic with "iterator already in use".

This is particularly problematic because a subsequent change will make
more iterators reusable (specifically TBI iterators), such that e.g.
`MVCCIncrementalIterator` will be creating two concurrent reusable
iterators and trigger this panic.

This patch supports multiple iterators by creating new cloned iterators
if the existing iterator is in use. This preserves the pinned engine
state for the cloned iterators too.

Release note: None

erikgrinaker commented:
This branch has been resubmitted in these PRs:

* #81193
* #81194
* #81195
* #81242

Queued PRs will be rebased onto `master` when possible.
